File: app/src/main/java/com/example/pam_project/features/lists/create/CreateListActivity.java
package com.example.pam_project.features.lists.create;

import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.widget.EditText;
import android.widget.Spinner;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;

import com.example.pam_project.R;
import com.example.pam_project.di.ApplicationContainer;
import com.example.pam_project.di.ApplicationContainerLocator;
import com.example.pam_project.features.categories.list.CategoryInformation;
import com.example.pam_project.features.categories.spinner.SpinnerActivity;
import com.example.pam_project.features.categories.spinner.SpinnerCategoryAdapter;
import com.example.pam_project.repositories.categories.CategoriesRepository;
import com.example.pam_project.repositories.lists.ListsRepository;

import java.util.List;
import java.util.Objects;

public class CreateListActivity extends AppCompatActivity implements CreateListView {
    private CreateListPresenter presenter;
    private SpinnerCategoryAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_create_list);
        createPresenter();
        Objects.requireNonNull(getSupportActionBar()).setTitle(R.string.activity_title_create_list);
        setUpView();
    }

    private void createPresenter() {
        final ApplicationContainer container = ApplicationContainerLocator.locateComponent(this);
        final CategoriesRepository categoriesRepository = container.getCategoriesRepository();
        final ListsRepository listsRepository = container.getListsRepository();
        presenter = new CreateListPresenter(categoriesRepository, listsRepository, this);
    }

    private void setUpView() {
        Spinner spinner = findViewById(R.id.create_list_category_spinner);
        adapter = new SpinnerCategoryAdapter(this, android.R.layout.simple_spinner_item);
        adapter.getCategories().setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        spinner.setAdapter(adapter.getCategories());
        SpinnerActivity spinnerActivity = new SpinnerActivity();
        spinner.setOnItemSelectedListener(spinnerActivity);
    }

    @Override
    public void onStart() {
        super.onStart();
        presenter.onViewAttached();
    }

    @Override
    public void bindCategories(final List<CategoryInformation> model) {
        adapter.update(model);
    }

    @Override
    public void onListInsertedError() {
        Toast.makeText(getApplicationContext(), getString(R.string.error_list_create), Toast.LENGTH_LONG).show();
    }

    @Override
    public void onCategoriesReceivedError() {
        Toast.makeText(getApplicationContext(), getString(R.string.error_categories_fetch), Toast.LENGTH_LONG).show();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.check_action_bar, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        long itemId = item.getItemId();
        final Spinner spinner = findViewById(R.id.create_list_category_spinner);
        final EditText listNameInput = findViewById(R.id.create_list_title_input);

        if (itemId == R.id.check_add_button) {
            String listName = listNameInput.getText().toString();
            Long categoryId = adapter.getCategoriesMap().get(spinner.getSelectedItem().toString());
            String errorMessage = checkForm(listName);

            if (errorMessage != null) {
                listNameInput.setError(errorMessage);
            } else {
                presenter.insertList(listName, categoryId);
                finish();
            }
        }

        return super.onOptionsItemSelected(item);
    }

    private String checkForm(String listName) {
        String errorMessage = null;

        if (listName == null || listName.trim().isEmpty()) {
            errorMessage = getString(R.string.error_empty_input);
        }

        return errorMessage;
    }

    @Override
    public void onStop() {
        super.onStop();
        presenter.onViewDetached();
    }
}

File: app/src/main/java/com/example/pam_project/features/lists/create/CreateListActivity.kt
package com.example.pam_project.features.lists.create

import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.widget.EditText
import android.widget.Spinner
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.example.pam_project.R
import com.example.pam_project.di.ApplicationContainerLocator
import com.example.pam_project.features.categories.list.CategoryInformation
import com.example.pam_project.features.categories.spinner.SpinnerActivity
import com.example.pam_project.features.categories.spinner.SpinnerCategoryAdapter
import com.example.pam_project.utils.validators.FormValidator
import java.util.*

class CreateListActivity : AppCompatActivity(), CreateListView {
    private lateinit var presenter: CreateListPresenter
    private lateinit var adapter: SpinnerCategoryAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_create_list)
        createPresenter()
        supportActionBar?.title=R.string.activity_title_create_list.toString()
        setUpView()
    }

    private fun createPresenter() {
        val container = ApplicationContainerLocator.locateComponent(this)
        val schedulerProvider = container.schedulerProvider
        val categoriesRepository = container.categoriesRepository
        val listsRepository = container.listsRepository
        presenter = CreateListPresenter(schedulerProvider, categoriesRepository,
                listsRepository, this)
    }

    private fun setUpView() {
        val spinner = findViewById<Spinner>(R.id.create_list_category_spinner)
        adapter = SpinnerCategoryAdapter(this, android.R.layout.simple_spinner_item)
        adapter.categories.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        spinner.adapter = adapter.categories
        val spinnerActivity = SpinnerActivity()
        spinner.onItemSelectedListener = spinnerActivity
    }

    public override fun onStart() {
        super.onStart()
        presenter.onViewAttached()
    }

    override fun bindCategories(model: List<CategoryInformation?>?) {
        adapter.update(model)
    }

    override fun onListInsertedError() {
        Toast.makeText(applicationContext, getString(R.string.error_list_create), Toast.LENGTH_LONG).show()
    }

    override fun onCategoriesReceivedError() {
        Toast.makeText(applicationContext, getString(R.string.error_categories_fetch), Toast.LENGTH_LONG).show()
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.check_action_bar, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val itemId = item.itemId.toLong()
        val spinner = findViewById<Spinner>(R.id.create_list_category_spinner)
        val listNameInput = findViewById<EditText>(R.id.create_list_title_input)
        if (itemId == R.id.check_add_button.toLong()) {
            val listName = listNameInput.text.toString()
            val categoryId = adapter.categoriesMap[spinner.selectedItem.toString()]
            val validForm = FormValidator.validate(applicationContext, createInputMap(listName, listNameInput))
            if (validForm) {
                presenter.insertList(listName, categoryId)
                finish()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    private fun createInputMap(listName: String, listNameInput: EditText): Map<EditText, String?> {
        val map: MutableMap<EditText, String?> = HashMap()
        map[listNameInput] = listName
        return map
    }

    public override fun onStop() {
        super.onStop()
        presenter.onViewDetached()
    }
}